### PR TITLE
Interval

### PR DESCRIPTION
**Pull request by ThijsWouters**
Is this something you would be interested in?

If so, I will develop it further.

### CODE REVIEW

**Comment by ThijsWouters** [comment]
I think that a struct is a way of naming things. And naming things is always good. That is what brought me to this library you named the datetime tuple.

I don't think an interval makes sense for `Date` and `Time` because they are not precise points in time, while `DateTime` is. 

The problem with an interval for `Date` is, where will you start and stop the interval. Do you start at midnight? Will you include the last date or not? And will you include it entirely?  

With `Time` it is even more difficult:
I have an interval of 12 hours until 13 hours. What is the duration of this interval? Is it 1 hour or is it 23?
What will you do with daylight saving time? We have an interval of 2:00 until 3:00. But we moved the clocks forward that day: so the duration is 0. But only on certain dates.

The functionality I am looking for is:
- calculating the duration of an interval, taking into account time-zones and dst,
- checking whether a `DateTime` is included in the interval,
- checking whether two intervals overlap (completely or partially).

**Comment by lau** [comment]
> I think that a struct is a way of naming things. And naming things is always good. That is what brought me to this library you named the datetime tuple.

You're right. Probably the functions could eventually also be made to use protocols so they could accept a tuple with two `DateTime`s which would be converted to an interval.

I think the functionality sounds good. The module name should be called `Calendar.DateTime.Interval`. Reasons explained below.

> The problem with an interval for Date is, where will you start and stop the interval. Do you start at midnight? Will you include the last date or not? And will you include it entirely? 

`Calendar.Date` has no concept of midnight and you can only include an entire date or not. The question of whether the end-date should be included or not is also something you have to decide when you have a DateTime, but just on a different scale. Is the last second included or not? It is just at second or microsecond granularity instead of at a day granularity. If you want an interval that ends at midnight will it be end 23:59:59.999999 or 00:00:00 the next day?

I think intervals should include both the beginning and the end.

With `Time` it makes less sense, which is why I didn't mention it. But for `NaiveDateTime` and `Date`, intervals can make sense too.

Yeah, it sounds interesting. :+1:

**Comment by ThijsWouters** [comment]
I also added types and specs. Is that something you want? It is in any case something that helps with static analysis.

File: lib/calendar/interval.ex [hunk 1]
```diff
@@ -0,0 +1,19 @@
+defmodule Calendar.Interval do
```
Comment (lau) [inline lib/calendar/interval.ex]: It should be called `Calendar.DateTime.Interval`

**Comment by lau** [comment]
Yeah, it's great that you added types and specs.

File: lib/calendar.ex [hunk 2]
```diff
@@ -12,6 +12,7 @@ defmodule Calendar do
       alias Calendar.TimeZoneData
       alias Calendar.TzPeriod
       alias Calendar.Strftime
+      alias Calendar.Interval
```
Comment (lau) [inline lib/calendar.ex]: Oh, sorry I hadn't seen your last commit and was just looking at this line, thinking it wasn't change in the module itself. Could you change this to `Calendar.DateTime.Interval` ? Then I think this PR will be good to go.

**Comment by ThijsWouters** [comment]
I'll resolve the conflicts, and then push again.

**Comment by lau** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**